### PR TITLE
Improve parser

### PR DESCRIPTION
**Pull request by YuriRomanowski**
## Description

Further parser improvements.
Changes:

1. Add parsing of year, allow arbitrary ordering of time reference parts;
2. Add more time formats (according to the [table](https://www.notion.so/serokell/Requirements-7dc47068a206498989c4b899a767d208) here;
3. Add date formats.

## Related issue(s)

Continuation of #1 

#### Related changes (conditional)

- Tests
  - [x] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [ ] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)
    - Haddock

#### Stylistic guide (mandatory)

- [x] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).

### CODE REVIEW

**Comment by dcastro** [comment]
Great work! :partying_face:  This actually fixes 2 issues I had noticed, but hadn't created a github issue yet:

* `01:30am in America/Winnipeg on 6th March` is parsed correctly, but if you reverse the order of the components, i.e. `01:30am on 6th March in America/Winnipeg`, then only the time and date were parsed. The timezone was ignored.
* `1am on 6th November 2022` - the year was being ignored

This ended up also fixing the issue @Martoon-00 described at the [bottom of this page](https://www.notion.so/serokell/Testing-iteration-1-from-dbfd12f048824fd68057f664fead248f), his example seems to work correctly now:

image

File: src/TzBot/TimeReference.hs [hunk 1]
```diff
@@ -49,7 +47,7 @@ data TimeReference = TimeReference
 data DateReference
   = DaysFromToday Int
   | DayOfWeekRef DayOfWeek
-  | DayOfMonthRef DayOfMonth (Maybe MonthOfYear)
+  | DayOfMonthRef DayOfMonth (Maybe MonthOfYear) (Maybe Year)
```
Comment (dcastro) [inline src/TzBot/TimeReference.hs:50]: You can't have a `Year` without also providing a `MonthOfYear` (e.g., "1am on the 5th 2022" is not an acceptable input), so maybe the definition should be more like this:

```suggestion
  | DayOfMonthRef DayOfMonth (Maybe (MonthOfYear (Maybe Year)))
```

Comment (YuriRomanowski) [reply src/TzBot/TimeReference.hs:50]: Yeah, I thought about it, your version is definitely more correct, but it was too ugly and I decided to leave current version :smile: But you are right, let's fix it eventually

File: src/TzBot/Parser.hs [hunk 2]
```diff
@@ -158,13 +159,52 @@ type TzParser = Parsec Void [Token]
...
+>>> parseTimeRefs "7:30pm 2022/08/3"
+[TimeReference {trText = "7:30pm 2022/08/3", trTimeOfDay = 19:30:00, trDateRef = Just (DayOfMonthRef 3 (Just (8,Just 2022))), trLocationRef = Nothing}]
+
+>>> parseTimeRefs "2022.8.03 7:30 pm "
+[TimeReference {trText = "2022.8.03 7:30 pm", trTimeOfDay = 19:30:00, trDateRef = Just (DayOfMonthRef 3 (Just (8,Just 2022))), trLocationRef = Nothing}]
+
+>>> parseTimeRefs "7:30pm 2022.8.03 America/Havana"
+[TimeReference {trText = "7:30pm 2022.8.03 America/Havana", trTimeOfDay = 19:30:00, trDateRef = Just (DayOfMonthRef 3 (Just (8,Just 2022))), trLocationRef = Just (TimeZoneRef America__Havana)}]
+
 -}
```
Comment (dcastro) [inline src/TzBot/Parser.hs:201]: While we're here, what do you think of also adding support for spaces in between offsets, e.g. `UTC +3`? Seems to be something users might write, and `timely` seems to support it too.

Comment (YuriRomanowski) [reply src/TzBot/Parser.hs:201]: Sure, let's support this too!

**Review by dcastro — APPROVED** [review]

**Pull request merged**